### PR TITLE
Fix hdfs gateway concurrent map writes

### LINKED ISSUE #13587: [hdfs-gateway] "fatal error: concurrent map writes" after updating to latest Minio
Running Minio in hdfs gateway mode crashes with `fatal error: concurrent map writes` when listing files in certain directories. And the client gets `mc: <ERROR> Unable to list folder. Server not initialized, please try again.`. It's consistent on a handful of directories but I haven't found a pattern among them. The issue seems to have started after [minio.RELEASE.2021-03-04T00-53-13Z](https://dl.min.io/server/minio/release/linux-amd64/archive/minio.RELEASE.2021-03-04T00-53-13Z) . This problem did not exist in earlier versions.

## Expected Behavior

`mc ls local/directory-x/` should list files for any/all directories.

## Current Behavior
`mc ls local/directory-x/` fails for certain directories.

## Possible Solution
I don't have a solution but maybe some helpful info. I tested a bunch of versions and the last working version was [minio.RELEASE.2021-03-04T00-53-13Z](https://dl.min.io/server/minio/release/linux-amd64/archive/minio.RELEASE.2021-03-04T00-53-13Z). Anything after that fails with the same error message.

## Steps to Reproduce (for bugs)
1. Run Minio server version `RELEASE.2021-03-04T00-53-13Z`.
2.  List files.
```
$ mc ls local/directory-1/
[2021-11-05 00:37:19 CDT]     0B subdir1/
[2021-11-05 00:37:19 CDT]     0B subdir2/
```
3. Update to Minio server version `RELEASE.2021-11-03T03-36-36Z`.
4. List files.
```
$ mc ls local/directory-1/
mc: <ERROR> Unable to list folder. Server not initialized, please try again.
```
5. Inspect Minio server logs:
```
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: fatal error: concurrent map writes
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: fatal error: concurrent map writes
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: goroutine 479 [running]:
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: runtime.throw({0x2d16a88, 0x2622acf})
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: runtime/panic.go:1198 +0x71 fp=0xc000033440 sp=0xc000033410 pc=0x435851
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: runtime.mapdelete_faststr(0x28e4740, 0xc000242ae0, {0xc000f0f800, 0xb7})
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: runtime/map_faststr.go:385 +0x135 fp=0xc0000334a0 sp=0xc000033440 pc=0x412f95
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/cmd/gateway/hdfs.(*hdfsObjects).ListObjects.func1({_, _}, {_, _}, {_, _})
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/cmd/gateway/hdfs/gateway-hdfs.go:455 +0x2b7 fp=0xc0000339d8 sp=0xc0000334a0 pc=0x2622757
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/cmd.listObjects.func2()
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/cmd/object-api-common.go:325 +0x136 fp=0xc000033f28 sp=0xc0000339d8 pc=0x23fcd36
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/internal/sync/errgroup.(*Group).Go.func1()
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: github.com/minio/minio/internal/sync/errgroup/errgroup.go:123 +0x1cb fp=0xc000033fe0 sp=0xc000033f28 pc=0xfcff2b
Nov 05 05:12:14 ip-10-10-10-10.prod.org1.comus-west-1.compute.internal minio[3503]: runtime.goexit()
```
HTTP trace logs didn't seem helpful.

## Context
I attempted to upgrade to the latest Minio server version and started having this issue. I ended up having to revert to an older version for now.

## Your Environment
* Operating System and version (`uname -a`):
`Linux hostname 4.14.173-137.229.amzn2.x86_64 #1 SMP Wed Apr 1 18:06:08 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

### PR DESCRIPTION
**Pull request by danochoa**
## Description
Resolves https://github.com/minio/minio/issues/13587

## How to test this PR?
Listing files is working as usual after this change.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — COMMENTED** [review]
Here is the correct fix 

```diff
diff --git a/cmd/gateway/hdfs/gateway-hdfs.go b/cmd/gateway/hdfs/gateway-hdfs.go
index ebb5f32bc..62130d444 100644
--- a/cmd/gateway/hdfs/gateway-hdfs.go
+++ b/cmd/gateway/hdfs/gateway-hdfs.go
@@ -28,6 +28,7 @@ import (
        "path"
        "sort"
        "strings"
+       "sync"
        "syscall"
        "time"
 
@@ -407,6 +408,7 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
 
 // ListObjects lists all blobs in HDFS bucket filtered by prefix.
 func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string, maxKeys int) (loi minio.ListObjectsInfo, err error) {
+       var mutex sync.Mutex
        fileInfos := make(map[string]os.FileInfo)
        targetPath := n.hdfsPathJoin(bucket, prefix)
 
@@ -430,6 +432,9 @@ func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, d
        }
 
        getObjectInfo := func(ctx context.Context, bucket, entry string) (minio.ObjectInfo, error) {
+               mutex.Lock()
+               defer mutex.Unlock()
+
                filePath := path.Clean(n.hdfsPathJoin(bucket, entry))
                fi, ok := fileInfos[filePath]
 
@@ -443,7 +448,6 @@ func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, d
                        }
 
                        fi, ok = fileInfos[filePath]
-
                        if !ok {
                                err = fmt.Errorf("could not get FileInfo for path '%s'", filePath)
                                return minio.ObjectInfo{}, hdfsToObjectErr(ctx, err, bucket, entry)
```

File: cmd/gateway/hdfs/gateway-hdfs.go [hunk 1]
```diff
@@ -407,6 +408,7 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
 
 // ListObjects lists all blobs in HDFS bucket filtered by prefix.
 func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string, maxKeys int) (loi minio.ListObjectsInfo, err error) {
+        var mutex sync.Mutex
```
Comment (harshavardhana) [inline cmd/gateway/hdfs/gateway-hdfs.go]:
```suggestion
       var mutex sync.Mutex
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
make sure to run `gofmt` on your code

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by danochoa** [comment]
@harshavardhana this seems to be failing in my local test environment

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
> @harshavardhana this seems to be failing in my local test environment

can you share the failure?

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by danochoa** [comment]
Minio gateway crashes in a loop on startup:
```
Nov 06 01:13:48  systemd[1]: Started MinIO.
Nov 06 01:13:48  minio[30838]: WARNING: MINIO_ACCESS_KEY and MINIO_SECRET_KEY are deprecated.
Nov 06 01:13:48  minio[30838]: Please use MINIO_ROOT_USER and MINIO_ROOT_PASSWORD
Nov 06 01:13:50  minio[30838]: panic: runtime error: invalid memory address or nil pointer dereference 
Nov 06 01:13:50  minio[30838]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x28856e9]
Nov 06 01:13:50  minio[30838]: goroutine 70 [running]:
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.(*NotificationSys).ServerInfo(0x0, 0x2, 0x1e670aa, 0xc000022080)
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/notification.go:1205 +0x29
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.GetPeerOnlineCount(0x4a052a8, 0xc0015beed0) 
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/notification.go:1291 +0x3f
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.nodeHealthMetricsPrometheus(0xc000e12120)
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/metrics.go:111 +0x26
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.(*minioCollector).Collect(0xc000dd6290, 0xc000e12120)
Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/metrics.go:101 +0xe5
Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:446 +0x12b
Nov 06 01:13:50  minio[30838]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:538 +0xe4d
Nov 06 01:13:50  systemd[1]: minio.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Nov 06 01:13:50  systemd[1]: Unit minio.service entered failed state.
Nov 06 01:13:50  systemd[1]: minio.service failed.
Nov 06 01:13:50  systemd[1]: minio.service holdoff time over, scheduling restart.
```

**Comment by harshavardhana** [comment]
> Minio gateway crashes in a loop on startup:
> 
> ```
> Nov 06 01:13:48  systemd[1]: Started MinIO.
> Nov 06 01:13:48  minio[30838]: WARNING: MINIO_ACCESS_KEY and MINIO_SECRET_KEY are deprecated.
> Nov 06 01:13:48  minio[30838]: Please use MINIO_ROOT_USER and MINIO_ROOT_PASSWORD
> Nov 06 01:13:50  minio[30838]: panic: runtime error: invalid memory address or nil pointer dereference 
> Nov 06 01:13:50  minio[30838]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x28856e9]
> Nov 06 01:13:50  minio[30838]: goroutine 70 [running]:
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.(*NotificationSys).ServerInfo(0x0, 0x2, 0x1e670aa, 0xc000022080)
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/notification.go:1205 +0x29
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.GetPeerOnlineCount(0x4a052a8, 0xc0015beed0) 
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/notification.go:1291 +0x3f
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.nodeHealthMetricsPrometheus(0xc000e12120)
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/metrics.go:111 +0x26
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd.(*minioCollector).Collect(0xc000dd6290, 0xc000e12120)
> Nov 06 01:13:50  minio[30838]: github.com/minio/minio/cmd/metrics.go:101 +0xe5
> Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
> Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:446 +0x12b
> Nov 06 01:13:50  minio[30838]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
> Nov 06 01:13:50  minio[30838]: github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:538 +0xe4d
> Nov 06 01:13:50  systemd[1]: minio.service: main process exited, code=exited, status=2/INVALIDARGUMENT
> Nov 06 01:13:50  systemd[1]: Unit minio.service entered failed state.
> Nov 06 01:13:50  systemd[1]: minio.service failed.
> Nov 06 01:13:50  systemd[1]: minio.service holdoff time over, scheduling restart.
> ```

This is something else.

**Comment by harshavardhana** [comment]
You should set prometheus metrics to `/minio/v2/metrics/cluster`

**Comment by danochoa** [comment]
I was wondering the same, currently rebuilding from latest to test without my recent pr changes.

**Comment by harshavardhana** [comment]
```diff
git diff
diff --git a/cmd/metrics.go b/cmd/metrics.go
index 1ec4000c8..66f30bad7 100644
--- a/cmd/metrics.go
+++ b/cmd/metrics.go
@@ -108,6 +108,10 @@ func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 }
 
 func nodeHealthMetricsPrometheus(ch chan<- prometheus.Metric) {
+       if globalIsGateway {
+               return
+       }
+
        nodesUp, nodesDown := GetPeerOnlineCount()
        ch <- prometheus.MustNewConstMetric(
                prometheus.NewDesc(
```

You need this as well.

**Comment by danochoa** [comment]
Test succeeded by building from commit 9735f3d8f2af30e00ac7162bd8d61fddad4b2521 with the patches in this PR applied. Building from latest (19bbf3e142d62b69500ce9432b9fe07c3e490190) failed with similar startup crash loop above, with and without patches applied. Also fails with 8774d10bdf25fc345e9deaf51131e3582a7f6e23.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**